parser/src/function.rs

```diff
@@ -385,4 +404,334 @@ impl<'input> InlinedFunction<'input> {
+    pub fn return_address(&self) -> Option<u64> {
+        self.return_address
+    }
+
+    /// The address of the call instruction.
+    #[inline]
+    pub fn called_from_address(&self) -> Option<&CalledFromAddress> {
+        self.called_from_address.as_ref()
+    }
+
+    /// The origin of the function being called.
+    #[inline]
+    pub fn origin(&self) -> Option<&FunctionCallOrigin<'input>> {
+        self.origin.as_ref()
+    }
+
+    /// The target location of the call.
+    #[inline]
+    pub fn target_locations(&self) -> &[(Range, Piece)] {
+        &self.target_locations
+    }
+
+    /// The registers where the parameter value is sent to.
+    #[inline]
+    pub fn target_registers(&self) -> impl Iterator<Item = (Range, Register)> + '_ {
+        location::registers(&self.target_locations)
+    }
+
+    /// The registers pointing to where this variable is stored.
+    pub fn target_register_offsets(&self) -> impl Iterator<Item = (Range, Register, i64)> + '_ {
+        location::register_offsets(&self.target_locations)
+    }
+
+    /// The stack frame locations at which this variable is stored.
+    pub fn target_frame_locations(&self) -> impl Iterator<Item = (Range, FrameLocation)> + '_ {
+        location::frame_locations(&self.target_locations)
+    }
+
+    /// Whether the target is clobbered.
+    #[inline]
+    pub fn target_is_clobbered(&self) -> bool {
+        self.target_is_clobbered
+    }
+
+    /// The called_function_type.
+    ///
+    /// Returns `None` if the called_function_type is invalid.
+    #[inline]
+    pub fn called_function_type<'a>(
+        &self,
+        hash: &'a FileHash<'input>,
+    ) -> Option<Cow<'a, Type<'input>>> {
+        if let Some(ty) = self.called_function_ty {
+            Type::from_offset(hash, ty)
+        } else {
+            None
+        }
+    }
+
+    /// The source location of the call.
+    #[inline]
+    pub fn called_from_source(&self) -> Option<&Source<'input>> {
+        self.called_from_source.as_ref()
+    }
+
+    /// The parameter inputs for this call.
+    #[inline]
+    pub fn parameter_inputs(&self) -> &[FunctionCallParameter<'input>] {
+        &self.parameter_inputs
+    }
+}
+
+/// The kind of function call being made.
+#[derive(Debug, Default, Clone, Copy, Eq, PartialEq, PartialOrd, Ord)]
+pub enum FunctionCallKind {
+    /// This is a normal function call made via a call-type instruction
+    #[default]
+    Normal,
+    /// This is a tail-call made via a jump-type instruction
+    Tail,
+}
+
+/// This is the function definition (origin) which is being called (if known)
+#[derive(Debug)]
+pub enum FunctionCallOrigin<'input> {
+    /// The static function definition which is being called directly.
+    Direct(&'input Function<'input>),
+    /// The indirect function definition
+    Indirect(FunctionCallIndirectOrigin<'input>),
+}
+
+/// Represents the subroutine pointer that is being called
+#[derive(Debug)]
+pub enum FunctionCallIndirectOrigin<'input> {
```

Comment:
This adds a lot of complexity to the API.

Do you need to know exactly which variable/parameter/member is the origin? Or is there some information in particular that you need which `ddbug-parser` could parse for you? For example, maybe you only need to know the type of the function being called?

gdb only uses this to get the address of the function being called (but it has the advantage that it can evaluate the target expression to determine the address if needed).

Have you encountered all of these variants in practice? I'd be interested in an easy way to reproduce DWARF for all of these for testing.